File: go/mechanisms/evm/batched/client/storage.go
package client

import (
	"sync"
)

// BatchedClientContext holds per-channel session state on the client side.
type BatchedClientContext struct {
	ChargedCumulativeAmount string `json:"chargedCumulativeAmount"`
	Balance                 string `json:"balance"`
	TotalClaimed            string `json:"totalClaimed"`
	DepositAmount           string `json:"depositAmount,omitempty"`
	SignedMaxClaimable      string `json:"signedMaxClaimable,omitempty"`
	Signature               string `json:"signature,omitempty"`
}

// ClientChannelStorage is the interface for persisting client-side channel sessions.
type ClientChannelStorage interface {
	Get(channelId string) (*BatchedClientContext, error)
	Set(channelId string, ctx *BatchedClientContext) error
	Delete(channelId string) error
}

// InMemoryClientChannelStorage is a volatile in-memory implementation of ClientChannelStorage.
type InMemoryClientChannelStorage struct {
	mu       sync.RWMutex
	sessions map[string]*BatchedClientContext
}

// NewInMemoryClientChannelStorage creates a new in-memory client session storage.
func NewInMemoryClientChannelStorage() *InMemoryClientChannelStorage {
	return &InMemoryClientChannelStorage{
		sessions: make(map[string]*BatchedClientContext),
	}
}

func (s *InMemoryClientChannelStorage) Get(channelId string) (*BatchedClientContext, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	ctx, ok := s.sessions[channelId]
	if !ok {
		return nil, nil
	}
	// Return a copy
	copy := *ctx
	return &copy, nil
}

func (s *InMemoryClientChannelStorage) Set(channelId string, ctx *BatchedClientContext) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	copy := *ctx
	s.sessions[channelId] = &copy
	return nil
}

func (s *InMemoryClientChannelStorage) Delete(channelId string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.sessions, channelId)
	return nil
}

File: python/x402/mechanisms/evm/batch_settlement/client/storage.py
"""Client-side channel storage protocol and in-memory default."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol


@dataclass
class BatchSettlementClientContext:
    """Client-side channel state for an open payment channel."""

    charged_cumulative_amount: str | None = None
    balance: str | None = None
    total_claimed: str | None = None
    signed_max_claimable: str | None = None
    signature: str | None = None

    # Free-form passthrough kept for forward-compat with future fields.
    extra: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = dict(self.extra)
        if self.charged_cumulative_amount is not None:
            out["chargedCumulativeAmount"] = self.charged_cumulative_amount
        if self.balance is not None:
            out["balance"] = self.balance
        if self.total_claimed is not None:
            out["totalClaimed"] = self.total_claimed
        if self.signed_max_claimable is not None:
            out["signedMaxClaimable"] = self.signed_max_claimable
        if self.signature is not None:
            out["signature"] = self.signature
        return out

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> BatchSettlementClientContext:
        known = {
            "chargedCumulativeAmount",
            "balance",
            "totalClaimed",
            "signedMaxClaimable",
            "signature",
        }
        return cls(
            charged_cumulative_amount=(
                str(data["chargedCumulativeAmount"])
                if data.get("chargedCumulativeAmount") is not None
                else None
            ),
            balance=str(data["balance"]) if data.get("balance") is not None else None,
            total_claimed=(
                str(data["totalClaimed"]) if data.get("totalClaimed") is not None else None
            ),
            signed_max_claimable=(
                str(data["signedMaxClaimable"])
                if data.get("signedMaxClaimable") is not None
                else None
            ),
            signature=data.get("signature"),
            extra={k: v for k, v in data.items() if k not in known},
        )

    def copy(self) -> BatchSettlementClientContext:
        return BatchSettlementClientContext(
            charged_cumulative_amount=self.charged_cumulative_amount,
            balance=self.balance,
            total_claimed=self.total_claimed,
            signed_max_claimable=self.signed_max_claimable,
            signature=self.signature,
            extra=dict(self.extra),
        )


class ClientChannelStorage(Protocol):
    """Sync client-side persistence for batch-settlement channels."""

    def get(self, key: str) -> BatchSettlementClientContext | None: ...
    def set(self, key: str, context: BatchSettlementClientContext) -> None: ...
    def delete(self, key: str) -> None: ...


class InMemoryClientChannelStorage:
    """Default in-process implementation of `ClientChannelStorage`."""

    def __init__(self) -> None:
        self._channels: dict[str, BatchSettlementClientContext] = {}

    def get(self, key: str) -> BatchSettlementClientContext | None:
        existing = self._channels.get(key)
        return existing.copy() if existing is not None else None

    def set(self, key: str, context: BatchSettlementClientContext) -> None:
        self._channels[key] = context.copy()

    def delete(self, key: str) -> None:
        self._channels.pop(key, None)


__all__ = [
    "BatchSettlementClientContext",
    "ClientChannelStorage",
    "InMemoryClientChannelStorage",
]
